test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/ClientTests/Tests/ClientEntityDescriptorTests.cs

```diff
@@ -0,0 +1,202 @@
+        {
+            var people = from p in _context.People
+                         orderby p.FirstName descending
+                         select p;
+
+            Assert.Equal("http://localhost/odata/People?$orderby=FirstName desc", people.ToString());
+
+            Assert.NotNull(people.FirstOrDefault());
+        }
+
+        [Fact]
+        public void GetDiscontinuedProducts()
+        {
+            var discontinuedProducts = from product in _context.Products
+                                       where product.Discontinued
+                                       select product;
+
+            Assert.Equal("http://localhost/odata/Products?$filter=Discontinued", discontinuedProducts.ToString());
+
+            Assert.NotNull(discontinuedProducts.FirstOrDefault());
+        }
+
+        [Fact]
+        public void GetInCirculationProducts()
+        {
+            var productsInCirculation = from product in _context.Products
+                                       where !product.Discontinued
+                                       select product;
+
+            Assert.Equal("http://localhost/odata/Products?$filter=not Discontinued", productsInCirculation.ToString());
+
+            Assert.NotNull(productsInCirculation.FirstOrDefault());
+        }
+
+        [Fact]
+        public void GetOrdersOnMyBirthday()
+        {
+            var dateTimeType = new DateTime?(DateTime.Now).GetType();
+            var ordersOnMyBirthday = from order in _context.Orders
+                                     where order.OrderDate.Day == 29 && order.OrderDate.Month == 5
+                                     select order;
+
+            Assert.Equal("http://localhost/odata/Orders?$filter=day(OrderDate) eq 29 and month(OrderDate) eq 5", ordersOnMyBirthday.ToString());
+
+            Assert.NotNull(ordersOnMyBirthday.FirstOrDefault());
+        }
+
+        [Fact]
+        public void GetCustomersInLondon()
+        {
+            var customersInLondon = from customer in _context.Customers
+                                    where customer.City == "London"
+                                    select customer;
+
+            Assert.Equal("http://localhost/odata/Customers?$filter=City eq 'London'", customersInLondon.ToString());
+
+            Assert.NotNull(customersInLondon.FirstOrDefault());
+        }
+
+        [Fact]
+        public void GetCustomersByKey()
+        {
+            IQueryable<Common.Client.Default.Customer> customerByKey = _context.Customers.Where(c => c.PersonID == 1);
+            Assert.Equal("Bob", customerByKey.FirstOrDefault().FirstName);
+        }
+
+        [Fact]
+        public void GetOrderDetailsByKey()
+        {
+            IQueryable<Common.Client.Default.OrderDetail> orderByKey = _context.OrderDetails.Where(c => c.OrderID == 7 && c.ProductID == 5);
+            Assert.Equal(7, orderByKey.FirstOrDefault().OrderID);
+            Assert.Equal(5, orderByKey.FirstOrDefault().ProductID);
+        }

```

Comment:
Since we have a separate `ByKey()`, I think should rename these functions to reflect they are filtering by key or doing a key lookup using the where clause. And since we have changed how `Where(c => c.PersonID == 1)` is translated by default in ODL 8, I think you should also assert the URL string like in the tests before.

Finally, I think you should also add test cases for the actually `ByKey()` method since it's now the recommended way to a retrieve an entity by ID.